tests/system/providers/google/bigquery/example_bigquery_dataset.py

```diff
@@ -0,0 +1,95 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+Example Airflow DAG for Google BigQuery service testing dataset operations.
+"""
+import os
+from datetime import datetime
+
+from airflow import models
+from airflow.operators.bash import BashOperator
+from airflow.providers.google.cloud.operators.bigquery import (
+    BigQueryCreateEmptyDatasetOperator,
+    BigQueryDeleteDatasetOperator,
+    BigQueryGetDatasetOperator,
+    BigQueryUpdateDatasetOperator,
+)
+from airflow.utils.trigger_rule import TriggerRule
+
+ENV_ID = os.environ.get("SYSTEM_TESTS_ENV_ID")
+DAG_ID = "bigquery_dataset"
+
+DATASET_NAME = f"dataset_{DAG_ID}_{ENV_ID}"
+
+
+with models.DAG(
+    DAG_ID,
+    schedule_interval="@once",
+    start_date=datetime(2021, 1, 1),
+    catchup=False,
+    tags=["example", "bigquery"],
+) as dag:
+    # [START howto_operator_bigquery_create_dataset]
+    create_dataset = BigQueryCreateEmptyDatasetOperator(task_id="create_dataset", dataset_id=DATASET_NAME)
+    # [END howto_operator_bigquery_create_dataset]
+
+    # [START howto_operator_bigquery_update_dataset]
+    update_dataset = BigQueryUpdateDatasetOperator(
+        task_id="update_dataset",
+        dataset_id=DATASET_NAME,
+        dataset_resource={"description": "Updated dataset"},
+    )
+    # [END howto_operator_bigquery_update_dataset]
+
+    # [START howto_operator_bigquery_get_dataset]
+    get_dataset = BigQueryGetDatasetOperator(task_id="get-dataset", dataset_id=DATASET_NAME)
+    # [END howto_operator_bigquery_get_dataset]
+
+    get_dataset_result = BashOperator(
+        task_id="get_dataset_result",
+        bash_command="echo \"{{ task_instance.xcom_pull('get-dataset')['id'] }}\"",
+    )
+
+    # [START howto_operator_bigquery_delete_dataset]
+    delete_dataset = BigQueryDeleteDatasetOperator(
+        task_id="delete_dataset", dataset_id=DATASET_NAME, delete_contents=True
+    )
+    # [END howto_operator_bigquery_delete_dataset]
+    delete_dataset.trigger_rule = TriggerRule.ALL_DONE
+
+    (
+        # TEST BODY
+        create_dataset
+        >> update_dataset
+        >> get_dataset
+        >> get_dataset_result
+        # TEST TEARDOWN
+        >> delete_dataset
+    )
+
+    from tests.system.utils.watcher import watcher
```

Comment:
Hmm. I thought a bit about our earlier discussion - https://github.com/apache/airflow/pull/22311#discussion_r829687607 with @ferruzzi and hmm. I changed my mind I think. 

 I think we have an easy way to decrease the amount of code here indeed in almost the way suggested :). And remove one of the "pre-commit checks".

```
from tests.system.utils.watcher import watcher

def get_test_run(dag, *, add_watcher: bool = False):
    def test_run():
        if add_watcher:
             # This test needs watcher in order to properly mark success/failure
             # when "tearDown" task with trigger rule is part of the DAG
              list(dag.tasks) >> watcher()
        dag.clear(dag_run_state=State.NONE)
        dag.run()

    return test_run
```

And then in the DAG:

```
test_run = get_test_run(dag, add_watcher=True)
```

This way we;

* do not need local import in the DAG
* do not need the pre-commit to add watcher code
* avoid code duplication

WDYT?

Reply:
I think this is brilliant 💎 
We have a `dag` reference there, so encapsulating everything in one place seems like a great idea. This makes code cleaner, and adding a watcher now only requires changing one flag. Seriously, collaborative thinking is amazing - many brains can come up with great ideas 😃 
I'm all for it.

Reply:
+1

Reply:
Ok, it was too soon to get excited before trying it out. After some tests I must say that it doesn't work as expected. 
When running tests using pytest, it reads the `get_test_run` function and it works fine, but when running them through the Airflow UI - it doesn't.
AFAIK The problem is that this function is outside of the DAG and Airflow UI doesn't read it, so the watcher is never added to the DAG. While when it is run with pytest, the watcher is there.
So I guess we are at the same spot where we were before this suggestion popped out.
I don't see any better way to simplify it, so I just want a confirmation from others, that we stick to what we have now, unless there is another idea. @potiuk @bhirsz

Reply:
I see. Right. Good reason. It was too beautiful. I though about addint a "get_watcher" method:

```
with DAG as dag:
      ....
      from utiils import add_watcher
      add_watcher(dag)
```

But apart of a litlle less code it does not change much.... so  guess we should stay with the original idea.

Reply:
> but when running them through the Airflow UI - it doesn't.

Am I understanding that to mean you are manually running tests from the Airflow UI?   I actually never realized that as a thing.   Or did you add that to a DASG which should have worked and it killed the DAG?

Reply:
Maybe not worth holding up the merge, but I wonder if moving the "add watcher" flag to the DAG class itself may be a resolution?  then usage might look like:

```
with DAG(
    dag_id='example_dag',
    schedule_interval=None,
    start_date=datetime(2021, 1, 1),
    tags=['example'],
    catchup=False,
    add_watcher=True,
) as dag:
```